fabric-tool-attributes-v1/src/main/java/net/fabricmc/fabric/api/tools/ToolAttributeHolder.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.tools;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+import net.minecraft.entity.EquipmentSlot;
+import net.minecraft.entity.attribute.EntityAttributeModifier;
+import net.minecraft.item.ItemStack;
+
+/**
+ * Interface for adding various tool attributes to items.
+ */
+public interface ToolAttributeHolder {
+	/**
+	 * @param stack The stack to check on.
+	 * @return The mining level of the item. 3 is equal to a diamond pick.
+	 */
+	int getMiningLevel(ItemStack stack);
+
+	/**
+	 * @param stack The stack to check on.
+	 * @return The mining speed of the item. 8.0 is equal to a diamond pick.
+	 */
+	float getMiningSpeed(ItemStack stack);
+
+	/**
+	 * Add modifiers for any {@link net.minecraft.entity.attribute.EntityAttributes} your item should give when equipped, based on the stack.
+	 * Appends to either attribute modifier NBT or the result from {@link net.minecraft.item.Item#getModifiers(EquipmentSlot)}.
+	 * If you do not want to append on a given item stack, add an NBT tag with the key "fabric_IgnoreDynamicModifiers".
+	 * @param slot The equipment slot this item is equipped in.
+	 * @param stack The stack that's equipped.
+	 * @return The dynamic modifiers to add on top of other modifiers on this stack.
+	 */
+	default Multimap<String, EntityAttributeModifier> getDynamicModifiers(EquipmentSlot slot, ItemStack stack) {
```

Comment:
I don't know how hot this method is. If it is, perhaps we should allow it to return null to avoid allocation of empty Multimaps.

Reply:
Is the returned map modifiable? If not you can have it return a static, immutable multimap with no entries.

Reply:
Multimaps are mutable.

Reply:
What I mean is: Are the maps returned by this method _intended_ to be mutated?

If there isn't an immutable multimap implemented everywhere, I'd be surprised, and I guess I'd agree that `null` would suffice. It's just having a non-null world would make life a lot safer in general.

Reply:
Actually, taking a look at my code again, the returned multimap from `getDynamicModifiers` is never mutated. I'm gonna add an EMPTY immutable multimap for folks to return if they don't want dynamic modifiers.